# Data Analysis Project on Weather Dataset

### Weather Dataset : 
#### The dataset used in this project is a time-series dataset with per-hour infromation about the weather conditions at a particular location.It records the wind speed,pressure,dew point temperature,visibility,relative humidity etc.

### Aim of the Project:
#### Exploratory Data Analysis will be performed on this dataset to look into some insights of the data which can help us in understanding weather conditions of this place better, so that this study can be used before making decisions in various fields such as agriculture,sport events,other events,business,travel etc.                                                                                                                       There is a set of 15 questios which will be answered on the basis of analysis.





## The commands to be used in this project are:

#### describe() - It is used to generate descriptive statistics that summarize the central tendency, dispersion and shape of a dataset’s distribution, excluding NaN values.

#### head() - It shows the first N rows in the data (by default, N=5).

#### unique() - In a column, it shows all the unique values. It can be applied on a single column only, not on the whole dataframe.

#### nunique() - It shows the total no. of unique values in each column. It can be applied on a single column as well as on the whole dataframe.

#### info() - Provides basic information about the dataframe.


## The attributes to be used in this project are:

#### shape - It shows the total no. of rows and no. of columns of the dataframe

#### index - This attribute provides the index of the dataframe

#### columns - It shows the name of each column

#### dtypes - It shows the data-type of each column

#### count - It shows the total no. of non-null values in each column. It can be applied on a single column as well as on the whole dataframe.

#### value_counts - In a column, it shows all the unique values with their count. It can be applied on a single column only.

### Importing pandas package as pd : 

In [3]:
import pandas as pd

In [38]:
weather=pd.read_csv(r"C:\Users\nafaz\Documents\\weatherdata.csv")
weather

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0        1/1/2012 0:00    -1.8              -3.9         86                4   
1        1/1/2012 1:00    -1.8              -3.7         87                4   
2        1/1/2012 2:00    -1.8              -3.4         89                7   
3        1/1/2012 3:00    -1.5              -3.2         88                6   
4        1/1/2012 4:00    -1.5              -3.3         88                7   
...                ...     ...               ...        ...              ...   
8779  12/31/2012 19:00     0.1              -2.7         81               30   
8780  12/31/2012 20:00     0.2              -2.4         83               24   
8781  12/31/2012 21:00    -0.5              -1.5         93               28   
8782  12/31/2012 22:00    -0.2              -1.8         89               28   
8783  12/31/2012 23:00     0.0              -2.1         86               30   

      Visibility_km  Press_kPa               Weather  
0               8.0     101.24                   Fog  
1               8.0     101.24                   Fog  
2               4.0     101.26  Freezing Drizzle,Fog  
3               4.0     101.27  Freezing Drizzle,Fog  
4               4.8     101.23                   Fog  
...             ...        ...                   ...  
8779            9.7     100.13                  Snow  
8780            9.7     100.03                  Snow  
8781            4.8      99.95                  Snow  
8782            9.7      99.91                  Snow  
8783           11.3      99.89                  Snow  

[8784 rows x 8 columns]

#### <> Information about the columns data in the dataset ;

In [5]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8784 entries, 0 to 8783
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date/Time         8784 non-null   object 
 1   Temp_C            8784 non-null   float64
 2   Dew Point Temp_C  8784 non-null   float64
 3   Rel Hum_%         8784 non-null   int64  
 4   Wind Speed_km/h   8784 non-null   int64  
 5   Visibility_km     8784 non-null   float64
 6   Press_kPa         8784 non-null   float64
 7   Weather           8784 non-null   object 
dtypes: float64(4), int64(2), object(2)
memory usage: 549.1+ KB


#### <> Getting total number of records and columns in the dataset ;

In [6]:
weather.shape

(8784, 8)

##### Total records/rows in the dataset = 8784
##### Total columns in the dataset = 8

#### <> Index start point,end point and step by which it increases;

In [7]:
weather.index

RangeIndex(start=0, stop=8784, step=1)

#### <> Name of the columns in the datset;

In [8]:
weather.columns

Index(['Date/Time', 'Temp_C', 'Dew Point Temp_C', 'Rel Hum_%',
       'Wind Speed_km/h', 'Visibility_km', 'Press_kPa', 'Weather'],
      dtype='object')

#### <> Datatypes of each column;

In [9]:
weather.dtypes

Date/Time            object
Temp_C              float64
Dew Point Temp_C    float64
Rel Hum_%             int64
Wind Speed_km/h       int64
Visibility_km       float64
Press_kPa           float64
Weather              object
dtype: object

#### <> Total number of items in each column;

In [10]:
weather.count()

Date/Time           8784
Temp_C              8784
Dew Point Temp_C    8784
Rel Hum_%           8784
Wind Speed_km/h     8784
Visibility_km       8784
Press_kPa           8784
Weather             8784
dtype: int64

##### As we can see, the count() function gives the number of non NULL values in columns, here we are getting equal number of values in each column that is 8784 so there is no null value in any column. 



#### <> Basic summary of each column;

In [11]:
weather.describe()

Temp_C  Dew Point Temp_C    Rel Hum_%  Wind Speed_km/h  \
count  8784.000000       8784.000000  8784.000000      8784.000000   
mean      8.798144          2.555294    67.431694        14.945469   
std      11.687883         10.883072    16.918881         8.688696   
min     -23.300000        -28.500000    18.000000         0.000000   
25%       0.100000         -5.900000    56.000000         9.000000   
50%       9.300000          3.300000    68.000000        13.000000   
75%      18.800000         11.800000    81.000000        20.000000   
max      33.000000         24.400000   100.000000        83.000000   

       Visibility_km    Press_kPa  
count    8784.000000  8784.000000  
mean       27.664447   101.051623  
std        12.622688     0.844005  
min         0.200000    97.520000  
25%        24.100000   100.560000  
50%        25.000000   101.070000  
75%        25.000000   101.590000  
max        48.300000   103.650000

##### As we have seen that the describe() method returns a kind of short summary of each column in the dataset.
##### Here is 25%,50%,70% represent First,Second and Third Quartile.
##### max and min represents maximum and minimum values in each column,std represents standard deviation,mean gives the average value of each column.

#### <> Retrieving first 10 records of the dataset;

In [12]:
weather.head(10)

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
0  1/1/2012 0:00    -1.8              -3.9         86                4   
1  1/1/2012 1:00    -1.8              -3.7         87                4   
2  1/1/2012 2:00    -1.8              -3.4         89                7   
3  1/1/2012 3:00    -1.5              -3.2         88                6   
4  1/1/2012 4:00    -1.5              -3.3         88                7   
5  1/1/2012 5:00    -1.4              -3.3         87                9   
6  1/1/2012 6:00    -1.5              -3.1         89                7   
7  1/1/2012 7:00    -1.4              -3.6         85                7   
8  1/1/2012 8:00    -1.4              -3.6         85                9   
9  1/1/2012 9:00    -1.3              -3.1         88               15   

   Visibility_km  Press_kPa               Weather  
0            8.0     101.24                   Fog  
1            8.0     101.24                   Fog  
2            4.0     101.26  Freezing Drizzle,Fog  
3            4.0     101.27  Freezing Drizzle,Fog  
4            4.8     101.23                   Fog  
5            6.4     101.27                   Fog  
6            6.4     101.29                   Fog  
7            8.0     101.26                   Fog  
8            8.0     101.23                   Fog  
9            4.0     101.20                   Fog

#### <> Number of unique records in each column;

In [13]:
weather.nunique()

Date/Time           8784
Temp_C               533
Dew Point Temp_C     489
Rel Hum_%             83
Wind Speed_km/h       34
Visibility_km         24
Press_kPa            518
Weather               50
dtype: int64

In [14]:
weather["Weather"].nunique()

50

###### There are total 50 distinct weather types in the data frame






#### <> All unique records in a particular column say Weather;

In [15]:
weather["Weather"].unique()

array(['Fog', 'Freezing Drizzle,Fog', 'Mostly Cloudy', 'Cloudy', 'Rain',
       'Rain Showers', 'Mainly Clear', 'Snow Showers', 'Snow', 'Clear',
       'Freezing Rain,Fog', 'Freezing Rain', 'Freezing Drizzle',
       'Rain,Snow', 'Moderate Snow', 'Freezing Drizzle,Snow',
       'Freezing Rain,Snow Grains', 'Snow,Blowing Snow', 'Freezing Fog',
       'Haze', 'Rain,Fog', 'Drizzle,Fog', 'Drizzle',
       'Freezing Drizzle,Haze', 'Freezing Rain,Haze', 'Snow,Haze',
       'Snow,Fog', 'Snow,Ice Pellets', 'Rain,Haze', 'Thunderstorms,Rain',
       'Thunderstorms,Rain Showers', 'Thunderstorms,Heavy Rain Showers',
       'Thunderstorms,Rain Showers,Fog', 'Thunderstorms',
       'Thunderstorms,Rain,Fog',
       'Thunderstorms,Moderate Rain Showers,Fog', 'Rain Showers,Fog',
       'Rain Showers,Snow Showers', 'Snow Pellets', 'Rain,Snow,Fog',
       'Moderate Rain,Fog', 'Freezing Rain,Ice Pellets,Fog',
       'Drizzle,Ice Pellets,Fog', 'Drizzle,Snow', 'Rain,Ice Pellets',
       'Drizzle,Snow,Fog', 

#### <> Number/Count of each unique record in the Weather column;

In [16]:
weather["Weather"].value_counts()

Mainly Clear                               2106
Mostly Cloudy                              2069
Cloudy                                     1728
Clear                                      1326
Snow                                        390
Rain                                        306
Rain Showers                                188
Fog                                         150
Rain,Fog                                    116
Drizzle,Fog                                  80
Snow Showers                                 60
Drizzle                                      41
Snow,Fog                                     37
Snow,Blowing Snow                            19
Rain,Snow                                    18
Thunderstorms,Rain Showers                   16
Haze                                         16
Drizzle,Snow,Fog                             15
Freezing Rain                                14
Freezing Drizzle,Snow                        11
Freezing Drizzle                        

In [1]:
weather["Temp_C"].max()

NameError: name 'weather' is not defined

In [18]:
#weather["Temp_C"].min()

### Following are the questions to be answered in this Project : 

### (Q1) Find all the unique 'Wind Speed' values in the data along with maximum,minimun and average wind speeds.

In [19]:
weather["Wind Speed_km/h"].unique()

array([ 4,  7,  6,  9, 15, 13, 20, 22, 19, 24, 30, 35, 39, 32, 33, 26, 44,
       43, 48, 37, 28, 17, 11,  0, 83, 70, 57, 46, 41, 52, 50, 63, 54,  2],
      dtype=int64)

In [20]:
max=weather["Wind Speed_km/h"].max()
print("Maximum wind speed : ",max," km/h")

Maximum wind speed :  83  km/h


In [21]:
min=weather["Wind Speed_km/h"].min()
print("Minimum wind speed : ",min," km/h")

Minimum wind speed :  0  km/h


In [22]:
avg=weather["Wind Speed_km/h"].mean()
print("Average wind speed : ",int(avg)," km/h")

Average wind speed :  14  km/h


### (Q2) Find the number of times when the 'Weather is exactly Clear'.

In [29]:
weather[weather.Weather=="Clear"].count()

Date/Time           1326
Temp_C              1326
Dew Point Temp_C    1326
Rel Hum_%           1326
Wind Speed_km/h     1326
Visibility_km       1326
Press_kPa           1326
Weather             1326
dtype: int64

###### Total number of instances of exactly clear weather : 1326

###### Note: value_counts() and groupby() method can also be used for this task. 

### (Q3) Find the number of times when the 'Wind Speed was exactly 4 km/h'.

In [41]:
weather[weather["Wind Speed_km/h"]==4].count()

Date/Time           474
Temp_C              474
Dew Point Temp_C    474
Rel Hum_%           474
Wind Speed_km/h     474
Visibility_km       474
Press_kPa           474
Weather             474
dtype: int64

###### Out of 8784 instances, 474 times the wind speed was recorded as 4 km/h

In [61]:
weather[4000:7000]

Date/Time  Temp_C  Dew Point Temp_C  Rel Hum_%  Wind Speed_km/h  \
4000   6/15/2012 16:00    25.0               5.1         28               15   
4001   6/15/2012 17:00    24.8               5.3         28               11   
4002   6/15/2012 18:00    24.5               6.0         30                9   
4003   6/15/2012 19:00    23.2               8.8         40               15   
4004   6/15/2012 20:00    22.0               8.7         43               15   
...                ...     ...               ...        ...              ...   
6995  10/18/2012 11:00    17.9               6.3         47               15   
6996  10/18/2012 12:00    18.3               6.3         45               17   
6997  10/18/2012 13:00    18.3               6.2         45               22   
6998  10/18/2012 14:00    19.3               5.8         41               28   
6999  10/18/2012 15:00    18.6               3.6         37               26   

      Visibility_km  Press_kPa        Weather  
4000           48.3     102.08         Cloudy  
4001           48.3     102.06         Cloudy  
4002           48.3     102.08         Cloudy  
4003           48.3     102.07         Cloudy  
4004           48.3     102.06  Mostly Cloudy  
...             ...        ...            ...  
6995           48.3     100.94   Mainly Clear  
6996           48.3     100.90   Mainly Clear  
6997           48.3     100.85   Mainly Clear  
6998           48.3     100.78   Mainly Clear  
6999           48.3     100.78  Mostly Cloudy  

[3000 rows x 8 columns]

### (Q4) Find out all the Null Values in the data.

In [59]:
weather[weather.isnull()].count()

Date/Time           0
Temp_C              0
Dew Point Temp_C    0
Rel Hum_%           0
Wind Speed_km/h     0
Visibility_km       0
Press_kPa           0
Weather             0
dtype: int64

###### There are no NULL values in the dataframe

### (Q5) Rename the column name 'Weather' of the dataframe to 'Weather Condition'.

In [ ]:
weather.

### (Q6) What is the mean 'Visibility' ?

In [60]:
weather["Visibility_km"].mean()

27.66444672131151

###### Mean Visibility is 27.63 km

### (Q7) What is the Standard Deviation of 'Pressure'  in this data?

In [7]:
Q. 1) Find all the unique 'Wind Speed' values in the data.
Q. 2) Find the number of times when the 'Weather is exactly Clear'.
Q. 3) Find the number of times when the 'Wind Speed was exactly 4 km/h'.
Q. 4) Find out all the Null Values in the data.
Q. 5) Rename the column name 'Weather' of the dataframe to 'Weather Condition'.
Q. 6) What is the mean 'Visibility' ?
Q. 7) What is the Standard Deviation of 'Pressure'  in this data?
Q. 8) What is the Variance of 'Relative Humidity' in this data ?
Q. 9) Find all instances when 'Snow' was recorded.
Q. 10) Find all instances when 'Wind Speed is above 24' and 'Visibility is 25'.
Q. 11) What is the Mean value of each column against each 'Weather Condition ?
Q. 12) What is the Minimum & Maximum value of each column against each 'Weather Condition ?
Q. 13) Show all the Records where Weather Condition is Fog.
Q. 14) Find all instances when 'Weather is Clear' or 'Visibility is above 40'.
Q. 15) Find all instances when :
A. 'Weather is Clear' and 'Relative Humidity is greater than 50'
or
B. 'Visibility is above 40'

SyntaxError: invalid syntax (Temp/ipykernel_18392/1438103274.py, line 1)

In [28]:
weather[weather.Weather=="Clear"].count()

Date/Time           1326
Temp_C              1326
Dew Point Temp_C    1326
Rel Hum_%           1326
Wind Speed_km/h     1326
Visibility_km       1326
Press_kPa           1326
Weather             1326
dtype: int64

In [58]:
weather[weather.isnull()].count()

Date/Time           0
Temp_C              0
Dew Point Temp_C    0
Rel Hum_%           0
Wind Speed_km/h     0
Visibility_km       0
Press_kPa           0
Weather             0
dtype: int64